### ISSUE TITLE
Receiving callback events starts a different react native instance

### ISSUE THREAD

**Issue by veedeo** [hoxfon/react-native-twilio-programmable-voice#12]
Every event that is received from the library, starts a different instance of React-Native activity. Is it possible to bring to front an original instance and forward events to it?

**Comment by fabriziomoscon** [comment]
@veedeo 
- What device/emulator are you using?
- Would it be possible to test your app changing this line:
https://github.com/hoxfon/react-native-twilio-programmable-voice/blob/master/android/src/main/java/com/hoxfon/react/TwilioVoice/NotificationHelper.java#L86

You can iterate quicker amending the code library from Android Studio, you can find the file browsing through your external libraries.
I made that change because I wanted the app to open up when not in memory, I haven't tested it on many devices, but surely there is space for improvement. Also I wanted the app to open up on when the screen is locked, but the implementation could be suboptimal.
If someone with more experience on Android development would read this, maybe they could advice.

Please let me know! Thanks for reporting

**Comment by fabriziomoscon** [comment]
@veedeo 
- what is the `android:launchMode` of your `MainActivity` specified in the `AndroidManifest.xml`?
I would suggest to have look at this https://developer.android.com/guide/topics/manifest/activity-element.html#lmode

I am trying different `Intent` flags and try to avoid the `Intent.FLAG_ACTIVITY_NEW_TASK`, will post an update

**Comment by fabriziomoscon** [comment]
@veedeo please check the new npm version:
react-native-twilio-programmable-voice@0.5.3

**Comment by veedeo** [comment]
You was right
I've added android:launchMode="singleTask" and now it works as expected
What was your fix in 0.5.3 that required a code change?

**Comment by fabriziomoscon** [comment]
I noticed that when the app is on the foreground (opened on the screen) there is no need to broadcast the intent for the received call, since the activity is already handling the intent.
As a result I had a double intent which made the phone ring twice, I guess that bug was somehow mitigated by `Intent.FLAG_ACTIVITY_NEW_TASK`.
Also I use `Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP` rather than `Intent.FLAG_ACTIVITY_NEW_TASK`

all improvements. I encourage you to update

**Comment by veedeo** [comment]
It crashes after update
```
Calling startActivity() from outside of an Activity  context requires the
FLAG_ACTIVITY_NEW_TASK flag. Is this really what you want?
startActivity
    ContextImpl.java:672
startActivity
    ContextImpl.java:659
startActivity
    ContextWrapper.java:331
startActivity
    ContextWrapper.java:331
onReactContextInitialized
    VoiceGCMListenerService.java:80
setupReactContext
    ReactInstanceManager.java:800
access$700
    ReactInstanceManager.java:104
onPostExecute
    ReactInstanceManager.java:228
onPostExecute
    ReactInstanceManager.java:197
finish
    AsyncTask.java:651
-wrap1
    AsyncTask.java:-1
handleMessage
    AsyncTask.java:668
dispatchMessage
    Handler.java:102
loop
    Looper.java:148
main
    ActivityThread.java:5417
invoke
    Method.java:-2
run
    ZygoteInit.java:726
main
    ZygoteInit.java:616

```

**Comment by veedeo** [comment]
It crashed after update
`Calling startActivity() from outside of an Activity  context requires the FLAG_ACTIVITY_NEW_TASK flag. Is this really what you want?
startActivity
    ContextImpl.java:672
startActivity
    ContextImpl.java:659
startActivity
    ContextWrapper.java:331
startActivity
    ContextWrapper.java:331
onReactContextInitialized
    VoiceGCMListenerService.java:80
setupReactContext
    ReactInstanceManager.java:800
access$700
    ReactInstanceManager.java:104
onPostExecute
    ReactInstanceManager.java:228
onPostExecute
    ReactInstanceManager.java:197
finish
    AsyncTask.java:651
-wrap1
    AsyncTask.java:-1
handleMessage
    AsyncTask.java:668
dispatchMessage
    Handler.java:102
loop
    Looper.java:148
main
    ActivityThread.java:5417
invoke
    Method.java:-2
run
    ZygoteInit.java:726
main
    ZygoteInit.java:616
`

**Comment by fabriziomoscon** [comment]
ok, I will set the Intent flag depending on the App Importance. Give me few minutes

**Comment by fabriziomoscon** [comment]
Alex, please try with this new version
`react-native-twilio-programmable-voice@0.5.4`
I have tested on a real device `ONEPlus 3`, which device/emulator are you testing on?

**Comment by veedeo** [comment]
Now the app just crashes without red screen
Android 6.0 APi level 23
I tried to revert to 0.5.2 and it works fine. the only problem is that it
duplicates events to react-native app
i recieve 4 events for deviceDidReceiveIncoming
and 2 events for connectionDidDisconnect

**Comment by fabriziomoscon** [comment]
I have tested only on Android 7.0

could you try to use `0.5.3` (which fixed the dubplications) and modify the `Intent` flag?
https://github.com/hoxfon/react-native-twilio-programmable-voice/blob/master/android/src/main/java/com/hoxfon/react/TwilioVoice/NotificationHelper.java#L93
The point is to find something that works in all these scenarios:
- app in foreground
- app in background
- app closed
- locked screen

**Comment by veedeo** [comment]
Ok I will check it, for now I just need to understand how to link it
directly to my RN and now via NPM

**Comment by fabriziomoscon** [comment]
I would just use AndroidStudio directly to amend the files. It is faster and. When you find a setting that doesn't crash you can report your findings. On the left > Project > External Libraries you can open and modify any java file. Then you just need to compile as you normally do.

**Comment by fabriziomoscon** [comment]
@veedeo I have released a new version:
react-native-twilio-programmable-voice@0.5.5
I believe all crashes are fixed and startActivity Intent flags used correctly. I have tested it in Android 6.0 and 7.0.
I have also made sure all wakeLock and guard and reset after I use them to wake up the screen from lock.
Please let me know if this fixes your crashes

**Comment by veedeo** [comment]
You are the man!
Now it works like a charm. However I noticed few things

1. when app was in memory (foreground/background) no calling notification in the status bar (this is a problem because you can hide ringing application, and will not see that its still ringing)
2. ringing notification with Dismiss/Answer buttons, can be closed by swiping, or by pressing Clear all on status bar. Then it will still ring but without any notification.

So to summarize, there always should be a Ringing notification and it should have FLAG_NO_CLEAR.

**Comment by fabriziomoscon** [comment]
Good points!
1. that is a feature, but becomes a bug in the scenario you described.
I noticed that it was not nice to have the notification when the app is on foreground. Maybe I could use `LifecycleEventListener`
https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/LifecycleEventListener.java
to create a notification when the app is put on background.

2. Yes well spotted. Once the ringtone was on the notification itself, but I found it easier to handle if it were a normal sound using play() and stop().

I think the best thing would be to close this issue and open 2 separate issues for 1 and 2 so I or other developers can look at them separately when they have a chance. I will try to look into them today.
